src/main/kotlin/com/demonwav/mcdev/platform/forge/creator/ForgeProjectCreator.kt

```diff
@@ -0,0 +1,317 @@
+        val dir = buildSystem.dirsOrError.resourceDirectory
+        runWriteTask {
+            CreatorStep.writeTextToFile(project, dir, ForgeConstants.MCMOD_INFO, text)
+        }
+    }
+}
+
+class Fg3ProjectFilesStep(
+    private val project: Project,
+    private val buildSystem: BuildSystem,
+    private val config: ForgeProjectConfig
+) : CreatorStep {
+
+    override fun runStep(indicator: ProgressIndicator) {
+        val modsTomlText = Fg3Template.applyModsToml(project, buildSystem, config)
+        val packMcmetaText = Fg3Template.applyPackMcmeta(project, buildSystem.artifactId)
+        val dir = buildSystem.dirsOrError.resourceDirectory
+        runWriteTask {
+            CreatorStep.writeTextToFile(project, dir, ForgeConstants.PACK_MCMETA, packMcmetaText)
+            val meta = dir.resolve("META-INF")
+            Files.createDirectories(meta)
+            CreatorStep.writeTextToFile(project, meta, ForgeConstants.MODS_TOML, modsTomlText)
+        }
+    }
+}
+
+class Fg3CompileJavaStep(
+    private val project: Project,
+    private val rootDirectory: Path
+) : CreatorStep {
+
+    override fun runStep(indicator: ProgressIndicator) {
+        indicator.text = "Setting up classpath"
+        indicator.text2 = "Running Gradle task: 'compileJava'"
+        runGradleTask(project, rootDirectory) { settings ->
+            settings.taskNames = listOf("compileJava")
+        }
+        indicator.text2 = null
+    }
+}
+
+enum class CreatedModuleType {
+    SINGLE, MULTI
+}
+
+class ForgeRunConfigsStep(
+    private val buildSystem: BuildSystem,
+    private val rootDirectory: Path,
+    private val config: ForgeProjectConfig,
+    private val createdModuleType: CreatedModuleType
+) : CreatorStep {
+
+    override fun runStep(indicator: ProgressIndicator) {
+        runWriteTaskLater {
+            val gradleDir = rootDirectory.resolve(".gradle")
+            Files.createDirectories(gradleDir)
+            val hello = gradleDir.resolve(HELLO)
+
+            val task = if (createdModuleType == CreatedModuleType.MULTI) {
+                ":${buildSystem.artifactId}:genIntellijRuns"
+            } else {
+                "genIntellijRuns"
+            }
+
+            // We don't use `rootModule.name` here because Gradle will change the name of the module to match
+            // what was set as the artifactId once it imports the project
+            val moduleName = if (createdModuleType == CreatedModuleType.MULTI) {
+                "${buildSystem.parentOrError.artifactId}.${buildSystem.artifactId}"
+            } else {
+                buildSystem.artifactId
+            }
+
+            val fileContents = moduleName + "\n" +
+                config.mcVersion + "\n" +
+                config.forgeVersion + "\n" +
+                task
+
+            CreatorStep.writeText(hello, fileContents)
+        }
+    }
+
+    companion object {
+        const val HELLO = ".hello_from_mcdev"
```

Comment:
No, this is an unholy hack because I've as of yet found no better solution.

The issue is twofold:

 1. When IntelliJ imports the Gradle project it will totally rework the project structure according to the Gradle project. Not even the `Module` objects will be valid anymore as it will create new modules
 2. There is no simple callback or event to listen for when Gradle has finished importing the project

So this is how I get around those 2 issues. IntelliJ's Gradle support does have a powerful mechanism for hooking into Gradle configurations and pulling information out - this is how I get the location of the SRG maps in FG right now. The issue is how do I know that a  Gradle import is the first complete import of a new FG project?

That's where this file comes in. This file gets written to `<project_dir>/.gradle` with some basic information regarding the project that was created. This way, when the code runs after a Gradle import I know I should set up those run configurations if that file exists, and the info in the file lets me both set up the run configurations correctly and handle false positivies.

This is a bit annoying but IntelliJ often imports the Gradle process multiple times before the project is fully formed,  even though I asked it not to. It doesn't cause any problems so I'm not worried about it, but that means I have to wait until the module that I'm expecting to exist actually exists.

You can see the code which creates that file where you commented [here](https://github.com/minecraft-dev/MinecraftDev/blob/feature/creator/src/main/kotlin/com/demonwav/mcdev/platform/forge/creator/ForgeProjectCreator.kt#L288-L311). But this is where it's actually used: [ForgeRunConfigDataService](https://github.com/minecraft-dev/MinecraftDev/blob/feature/creator/src/main/kotlin/com/demonwav/mcdev/platform/forge/gradle/ForgeRunConfigDataService.kt)